liblwgeom/lwgeom_geos.c

```diff
@@ -617,37 +620,47 @@ lwgeom_intersection(const LWGEOM *geom1, const LWGEOM *geom2)
 
 	LWDEBUG(3, "intersection() START");
 
-	g1 = LWGEOM2GEOS(geom1, 0);
-	if ( 0 == g1 )   /* exception thrown at construction */
+	g1 = LWGEOM2GEOS(geom1, 1);
+	if (!g1) /* exception thrown at construction */
 	{
 		lwerror("First argument geometry could not be converted to GEOS: %s", lwgeom_geos_errmsg);
-		return NULL ;
+		return NULL;
 	}
 
-	g2 = LWGEOM2GEOS(geom2, 0);
-	if ( 0 == g2 )   /* exception thrown at construction */
+	g2 = LWGEOM2GEOS(geom2, 1);
+	if (!g2) /* exception thrown at construction */
 	{
-		lwerror("Second argument geometry could not be converted to GEOS.");
+		lwerror("Second argument geometry could not be converted to GEOS: %s", lwgeom_geos_errmsg);
 		GEOSGeom_destroy(g1);
-		return NULL ;
+		return NULL;
 	}
 
 	LWDEBUG(3, " constructed geometrys - calling geos");
 	LWDEBUGF(3, " g1 = %s", GEOSGeomToWKT(g1));
 	LWDEBUGF(3, " g2 = %s", GEOSGeomToWKT(g2));
-	/*LWDEBUGF(3, "g2 is valid = %i",GEOSisvalid(g2)); */
-	/*LWDEBUGF(3, "g1 is valid = %i",GEOSisvalid(g1)); */
 
-	g3 = GEOSIntersection(g1,g2);
+	g3 = GEOSIntersection(g1, g2);
 
-	LWDEBUG(3, " intersection finished");
+	LWDEBUG(3, "intersection finished");
 
-	if (g3 == NULL)
+	if (!g3)
 	{
+		/* we're failing but not losing hope, try make all input valid */
+		if (!GEOSisValid(g1) || !GEOSisValid(g2))
+		{
+			g1 = LWGEOM_GEOS_makeValid(g1);
```

Comment:
We should fail on invalid inputs, not silently run a magic function that may or may not produce a desired results.